core/priority/prioritiser.go

```diff
@@ -0,0 +1,279 @@
+		for _, msg := range msgs[slot] {
+			slotMsgs = append(slotMsgs, msg)
+		}
+		result, err := calculateResult(slotMsgs, n.minRequired)
+		if err != nil {
+			log.Error(ctx, "Validate priority consensus", err) // Unexpected
+			return
+		}
+
+		err = n.consensus.Propose(ctx, slot, result)
+		if err != nil {
+			log.Warn(ctx, "Propose priority consensus", err) // Unexpected
+			return
+		}
+
+		completedSlot = slot
+		delete(msgs, slot)
+		delete(timeouts, slot)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case slot := <-n.trigger:
+			if slot <= completedSlot {
+				continue // Ignore triggers for completed slots.
+			}
+
+			err := n.prioritiseOnce(ctx, slot)
+			if err != nil {
+				log.Warn(ctx, "Priority error", err)
+			}
+
+		case msg := <-n.received:
+			if msg.Slot <= completedSlot {
+				continue // Ignore messages for completed slots.
+			}
+
+			slotMsgs, ok := msgs[msg.Slot]
+			if !ok {
+				slotMsgs = make(map[peer.ID]*pbv1.PriorityMsg)
+				timeouts[msg.Slot] = time.Now().Add(n.consensusTimeout)
+			}
+			slotMsgs[peer.ID(msg.PeerId)] = msg
+			msgs[msg.Slot] = slotMsgs
+
+			if len(slotMsgs) == len(n.peers) {
+				// All messages received before timeout
+				startConsensus(msg.Slot)
+			}
+		case now := <-ticker:
+			for slot, timeout := range timeouts {
+				if now.Before(timeout) {
+					continue
+				}
+
+				startConsensus(slot) // Note that iterating and deleting from a map from a single goroutine is fine.
+			}
+		}
+	}
+}
+
+// handleRequest handles a priority message exchange initiated by a peer.
+func (n *Prioritiser) handleRequest(ctx context.Context, pID peer.ID, msg *pbv1.PriorityMsg) (*pbv1.PriorityMsg, error) {
+	if string(pID) != msg.PeerId {
+		return nil, errors.New("invalid priority message peer id")
+	} else if err := n.msgValidator(msg); err != nil {
+		return nil, errors.Wrap(err, "invalid priority message")
+	}
+
+	select {
+	case n.received <- msg:
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-n.quit:
+		return nil, errors.New("prioritiser shutdown")
+	}
+
+	return n.msgProvider(msg.Slot)
+}
+
+// prioritiseOnce initiates a priority message exchange with all peers.
+func (n *Prioritiser) prioritiseOnce(ctx context.Context, slot int64) error {
+	msg, err := n.msgProvider(slot)
+	if err != nil {
+		return err
+	}
+
+	// Send our own message first to start consensus timeout.
+	n.received <- msg
+
+	for _, pID := range n.peers {
+		go func(pID peer.ID) {
+			response := new(pbv1.PriorityMsg)
+			ok := n.transport.SendReceive(ctx, pID, msg, response, protocolID)
+			if !ok {
+				// No need to log, since transport will do it.
+				return
+			}
+
+			if string(pID) != msg.PeerId {
+				log.Warn(ctx, "Invalid priority message peer id", nil)
+				return
+			}
+
+			if err := n.msgValidator(msg); err != nil {
+				log.Warn(ctx, "Invalid priority message from peer", err, z.Str("peer", p2p.PeerName(peer.ID(msg.PeerId))))
+				return
+			}
+
+			n.received <- msg
```

Comment:
```suggestion
			select {
	                case n.received <- msg:
	                case <-ctx.Done():
		               return
	                case <-n.quit:
		               return
	                }
```

Reply:
So that pattern is for external/other goroutines, not for when it is the same goroutine sending something to itself.